json-format.md

```diff
@@ -136,40 +130,45 @@ the [type system mapping](#22-type-system-mapping).
-
-If the `datacontenttype` value is either ["application/json"][rfc4627] or any
-media type with a [structured +json suffix][rfc6839], the implementation MUST
-translate the `data` attribute value into a [JSON value][json-value], and set
-the `data` attribute of the envelope JSON object to this JSON value.
-
-If the `datacontenttype` value does not follow the [structured +json
-suffix][rfc6839] but is known to use JSON encoding, the implementation MUST
-translate the `data` attribute value into a [JSON value][json-value], and set
-the `data` attribute of the envelope JSON object to this JSON value. Its typical
-examples are, but not limited to, `text/json`,
-[`application/json-seq`][json-seq] and
-[`application/geo+json-seq`][json-geoseq].
-
-Unlike all other attributes, for which value types are restricted to strings per
-the [type-system mapping](#22-type-system-mapping), the resulting `data` member
-[JSON value][json-value] is unrestricted, and MAY also contain numeric and
-logical JSON types.
-
-Second, whether a Base64-encoded string in the data attribute is treated as
-`Binary` or as a `String` is also determined by the `datacontenttype` value. If
-the `datacontenttype` media type is known to contain text, the data attribute
-value is not further interpreted and treated as a text string. Otherwise, it is
-decoded and treated as a binary value.
+The `data` attribute has no specified type, and so is treated specially,
+according to the following rules. In cases where multiple rules apply to the
+same data, the top most rule MUST be used.
+
+If the data attribute is known by the implementation to be JSON, the
+implementation MUST translate the `data` attribute value into a
+[JSON value][json-value], and set the `data` attribute of the envelope JSON
+object to this JSON value. An implementation MUST consider events with the
+following `datacontenttype` values as JSON:
+
+* [`"application/json"`][rfc4627]
+* Any value with a [structured `+json` suffix][rfc6839]
+
+An implementation MAY use other means to know if an event is representable as
+JSON, for example, using other well known `datacontentype` values such as
+`text/json` or `application/json-seq`][json-seq], or through inspection of
+types passed to an API.
+
+If the top-level type of `datacontenttype` is `text` (eg, `text/plain`), the
+implementation MUST translate the `data` attribute into a
+[JSON string][json-string] (taking the `datacontentencoding` attribute into
+consideration if the value is originally binary) and set the `data` attribute
+of the envelope JSON object to this JSON string.
+
+If the `datacontenttype` is anything else, or is not set, the implementation
+MUST translate the value to a [Base64-encoded][base64] 
+[JSON string][json-string], and set the `data` attribute of the envelope JSON
+object to this JSON string.
+
+Note that even if an implementation knows that the type of the `data` attribute
```

Comment:
Some of the original text here should have taken the introduction of `datacontentencoding` into account when that was introduced. A String is a binary if and only if `datacontentencoding` says so. Otherwise it's not. So this particular clause is already covered.

Reply:
I'm a bit confused actually about the `datacontentencoding` attribute. So, let's say I (an application developer writing a program that emits events) produce an event that is `application/x-csh` by passing a string of bytes as the data attribute value to the cloud events implementation that I'm using and setting the `datacontenttype` to `application/x-csh`. The implementation I'm working with has no idea what `application/x-csh`, it just treats it as a binary blob, like it would `application/octet-stream`, and sends it on to the next producer. But then it gets passed to another implementation, and that implementation knows that that's a c-shell script, which is text. But there's no `datacontentencoding` set, because the previous implementation that didn't know that it was text and so didn't know to ensure that there was a `datacontentencoding` attribute didn't set one.. So, what should it do? Treat it as text? Or as binary?

Another thing, how about `application/xml`? XML actually has an embedded charset parameter (ie `<?xml version="1.0" charset="utf-8"?>`). Should the `datacontentencoding` be set in that case? Are we therefore requiring all CloudEvents implementations, when they receive XML, to parse the XML, extract the charset, and set the `datacontentencoding` to it? I would have thought not, I would have thought they can just pass the binary blob as is without the `datacontentencoding`.

Anyway, this clause is not just redundant for the reason you stated, but it's also redundant because the behavior for when something should be `String` or `Binary` is completely specified above it. The reason I included it is because I can imagine many implementors making the mistake of encoding something like `application/xml` as a `String` when it should be base64ed `Binary`. Afterall, xml is text, so it's intuitive to handle it as text. So I thought this would be a good place to add that reminder to ensure that mistake isn't made.